contracts/governance/Badger.sol

```diff
@@ -0,0 +1,342 @@
+//SPDX-License-Identifier: Unlicense
+pragma solidity 0.8.10;
+
+import "@openzeppelin/contracts/token/ERC1155/ERC1155.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "../libraries/Strings.sol";
+
+contract Badger is Ownable, ERC1155 {
+    using Strings for string;
+
+    /*
+        Structs
+    */
+
+    struct TokenTier {
+        string uriId;
+        bool transferable;
+    }
+
+    /*
+        State variables
+    */
+
+    mapping(uint256 => TokenTier) public tokenTiers; // tokenId => TokenTier
+
+    /*
+        Modifiers
+    */
+
+    modifier isSameLength(
+        address[] calldata accounts,
+        uint256[] calldata tokenIds,
+        uint256[] memory amounts
+    ) {
+        require(
+            accounts.length == tokenIds.length &&
+                tokenIds.length == amounts.length,
+            "Input array mismatch"
+        );
+        _;
+    }
+
+    modifier isTier(uint256 tokenId) {
+        require(
+            _isNonEmptyString(tokenTiers[tokenId].uriId),
+            "Tier does not exist"
+        );
+        _;
+    }
+
+    modifier isValidString(string calldata uriId) {
+        require(_isNonEmptyString(uriId), "URI identifier required");
+        _;
+    }
+
+    modifier isValidTransfer(uint256 tokenId, address from) {
+        require(
+            tokenTiers[tokenId].transferable,
+            "Transfer disabled for this tier"
+        );
+        require(
+            msg.sender == owner() ||
+                from == _msgSender() ||
```

Comment:
I couldn't find where `_msgSender()` is coming from or what it does, altho I can guess. Is it just the same as `msg.sender`?

Reply:
I changed it to use `msgSender()` consistently.

Reply:
Thanks for the explanation